add 4pop HomSap ooa model from Jouganous et al 2017

rwaples opened the pull request:
I have attempted to implement the 4 population model presented in Jouganous et al 2017. 

This is my first time contributing to stdpopsim.

jeromekelleher commented:
Hi @rwaples, welcome to popsim! The model looks good to me, thanks very much for contributing. I think we should be able to merge this one pretty quickly, and then hopefully find someone to do the QC model. However, @grahamgower is more up-to-date on the protocols for adding new models and we should probably wait for his :+1: (I think he's back next week)

rwaples commented:
Thanks, sounds good! 

In regards to the model, I wasn't quite sure how to deal with population sizes prior to when the populations should exist, so I set them to zero, but looking at other model I see that this isn't always done.

jeromekelleher commented:
> In regards to the model, I wasn't quite sure how to deal with population sizes prior to when the populations should exist, so I set them to zero, but looking at other model I see that this isn't always done.

It's optional - once there's no lineages in a particular population, it doesn't really matter what its size is.

apragsdale commented on stdpopsim/catalog/HomSap/__init__.py:
```diff
@@ -1650,3 +1650,175 @@ def _AJ():
+        # JPT splits from CHB
+        msprime.MassMigration(time=T_CH_JP, source=3, destination=2, proportion=1.0),
+        # set growth rate in JPT to zero
+        msprime.PopulationParametersChange(
```
This parameters change shouldn't be necessary, as no lineages remain in the JPT population.

rwaples replied:
Yes, that makes sense.  I was struggling a bit how to deal with pops before they should exist.

apragsdale commented on stdpopsim/catalog/HomSap/__init__.py:
```diff
@@ -1650,3 +1650,175 @@ def _AJ():
+        msprime.MassMigration(time=T_EU_AS, source=2, destination=1, proportion=1.0),
+        # change size of B to N_B
+        # set growth rates in B and CHB to zero,
+        msprime.PopulationParametersChange(
```
Here, we probably don't want to be adding 0.0001 to the time. The demographic events in msprime are applied in the order that they are specified, so keeping the events in the current order and setting `time=T_EU_AS` should do it.

apragsdale commented on stdpopsim/catalog/HomSap/__init__.py:
```diff
@@ -1650,3 +1650,175 @@ def _AJ():
+        msprime.PopulationParametersChange(
+            time=T_EU_AS + 0.0001, initial_size=N_B, growth_rate=0.0, population_id=1
+        ),
+        msprime.PopulationParametersChange(
```
Also not needed - doesn't change the behavior of the model, but might end up causing the strict equality tests in QC to fail.

rwaples replied:
Yes, that makes sense.

apragsdale commented on stdpopsim/catalog/HomSap/__init__.py:
```diff
@@ -1650,3 +1650,175 @@ def _AJ():
+            time=T_EU_AS + 0.0001, growth_rate=0.0, population_id=2
+        ),
+        # set up migration between YRI and B
+        msprime.MigrationRateChange(
```
Here, I think we also want to avoid adding the 0.0001 buffer to the time of the migration rate changes.

rwaples replied:
I will remove this and the other 0.0001 buffers.  It is nice they are not necessary.

apragsdale commented on stdpopsim/catalog/HomSap/__init__.py:
```diff
@@ -1650,3 +1650,175 @@ def _AJ():
+    ]
+
+    # Change size of YRI/African population
+    AF_event = [
```
I think this should have `time=T_AF` instead of `T_B`.

rwaples replied:
Good catch!

apragsdale reviewed (changes requested):
Thanks for the contribution, @rwaples! This looks great. I left a few comments here. Like @jeromekelleher said, resetting parameters of populations with no lineages doesn't change the behavior, but might end up causing confusion during QC. Also, adding the timing buffer of 0.0001 to some events shouldn't be needed, since events are applied in the order (backwards in time) that they are specified (unless this causes some issue with the slim engine?).

Thanks again, and welcome to stdpopsim!!

rwaples commented:
Thanks for the helpful comments, you folks have really made it easy to contribute here. As you both noticed I did have a bit of confusion in how to deal with populations before they really exist.  

I have now removed the 0.0001 time buffers and also removed the demographic events that set pop sizes to zero when no lineages should exist in the population.  I will resubmit the pull request soon. 

I do plan to use this model (and others from stdpopsim) with the SLiM engine, via slim_script=True and with a bit of  subsequent editing of the SLiM script file.  I would be happy to help/test with that process if I can.

grahamgower reviewed (approved):
Thanks @rwaples, this looks excellent! Would you be able to squash your commits into a single commit following the dev docs. I think this is ready for merging after that.

rwaples commented:
Great.  I tried to squash into a single commit

grahamgower commented:
Merged. Thanks again @rwaples! And please do let me know if you have any issues with modifying/using the script output by the SLiM engine (I trust you've seen the [`recap_and_rescale()`](https://stdpopsim.readthedocs.io/en/latest/api.html#stdpopsim.slim_engine._SLiMEngine.recap_and_rescale) function?).

rwaples commented:
Thanks!
I have seen recap_and_rescale(), but I didn't end up using that as I am pretty familiar with simulate() and didn't do any rescaling, nice to have though.  I'm just using msprime.simulate(from_ts=).  

The only real issue I have found with stdpopsim -> SLiM so far is that an msprime.CensusEvent() within the demographic events does not seem to have any impact when running with the SLiM engine.  I have ended up manually editing the SLiM code to add "treeSeqRememberIndividuals".  I can raise an issue about this if it would help., but I also understand if census events are not really meant to be supported. 

FYI, this same paper has a related 5-pop model as well, so I will implement that once I get a chance.

grahamgower commented:
> The only real issue I have found with stdpopsim -> SLiM so far is that an msprime.CensusEvent() within the demographic events does not seem to have any impact when running with the SLiM engine. I have ended up manually editing the SLiM code to add "treeSeqRememberIndividuals". I can raise an issue about this if it would help., but I also understand if census events are not really meant to be supported.

I don't see any reason why we couldn't support `msprime.CensusEvent()`, so feel free to open an issue for that.


> FYI, this same paper has a related 5-pop model as well, so I will implement that once I get a chance.

:+1:

rwaples commented:
Quick question - There is an error in the description of this model (docs/parameter_tables/HomSap/OutOfAfrica_4J17.csv) (as noted in #757).  I can fix this, but not sure how to proceed - should I do this with a new pull request?

petrelharp commented:
Yes - a PR is the way to go. Thanks!
